aic-emu/GfxHandler.cc

```diff
@@ -0,0 +1,542 @@
+        return GFX_FAIL;
+
+
+    // -- Get FD
+    struct drm_prime_handle gem_handle_fd;
+    memset(&gem_handle_fd, 0, sizeof(gem_handle_fd));
+
+    gem_handle_fd.fd     = -1;
+    gem_handle_fd.handle = gem_create.handle;
+    gem_handle_fd.flags  = DRM_CLOEXEC | DRM_RDWR;
+
+    sts = RunIoctl(DRM_IOCTL_PRIME_HANDLE_TO_FD, &gem_handle_fd);
+
+    errString = "Ioctl PRIME_HANDLE_TO_FD: handle = " + std::to_string(gem_handle_fd.handle);
+    GFX_LOG_FAILURE(sts != 0, errString.c_str());
+    if (!GFX_SUCCESS(sts))
+        return GFX_FAIL;
+
+    // -- Insert into vector, referencing with remote handle specified in the YML log
+    boData_t boData = {.surf      = surf,
+                       .fd        = gem_handle_fd.fd,
+                       .handle    = gem_create.handle,
+                       .mmap_addr = nullptr};
+
+    m_boList.insert(std::make_pair(remote_handle, boData));
+
+#if (__DEBUG)
+    printf("%s: remote_handle = 0x%p .. prime_fd = %d\n",
+           __FUNCTION__, remote_handle, boData.fd);
+#endif
+
+    return GFX_OK;
+}
+
+GfxStatus GfxHandler::ClearBo(uint64_t remote_handle)
+{
+    GfxStatus sts = GFX_OK;
+    std::string errString;
+
+    auto bo_entry = m_boList.find(remote_handle);
+    if (bo_entry == m_boList.end())
+    {
+        std::cout << "Could not find bo with remote_handle= " << remote_handle << std::endl;
+        return GFX_FAIL;
+    }
+
+    struct drm_gem_close gem_close;
+    memset(&gem_close, 0, sizeof(gem_close));
+    gem_close.handle = bo_entry->second.handle;
+
+    sts = RunIoctl(DRM_IOCTL_GEM_CLOSE, &gem_close);
+
+    errString = "Ioctl GEM_CLOSE: handle = " + std::to_string(gem_close.handle);
+    GFX_LOG_FAILURE(sts != 0, errString.c_str());
+    if (!GFX_SUCCESS(sts))
+        return GFX_FAIL;
+
+    //Remove record from map
+    m_boList.erase(bo_entry);
+
+    return GFX_OK;
+}
+
+boData_t* GfxHandler::GetBo(uint64_t handle)
+{
+    if (m_boList.size() == 0)
+        return nullptr;
+
+    auto bo_entry = m_boList.find(handle);
+    if (bo_entry == m_boList.end())
+    {
+        std::cout << "Could not find bo with handle= " << handle << std::endl;
+        return nullptr;
+    }
+
+    return &bo_entry->second;
+}
+
+GfxStatus GfxHandler::Copy(boData_t* dstData, const void* srcBuf)
+{
+    if (!srcBuf || !dstData)
+    {
+        std::cout << "Error: Unexpected Null Ptr: srcBuf = 0x"<< std::hex << srcBuf
+                  << " , dstData = 0x" << std::hex << dstData << std::dec << std::endl;
+        return GFX_FAIL;
+    }
+
+    GfxStatus sts = GFX_OK;
+    std::string errString;
+
+    SurfaceParams_t surf = dstData->surf;
+    size_t sizeToWrite = surf.width * surf.height * surf.pixelSize;
+
+    //Map buffer for CPU write
+    sts = MmapBo(dstData, true);
+
+    errString = "MmapBo call: ";
+    GFX_LOG_FAILURE(sts != 0, errString.c_str());
+    if (!GFX_SUCCESS(sts))
+        return GFX_FAIL;
+
+    memcpy(dstData->mmap_addr, srcBuf, sizeToWrite);
```

Comment:
this does not seem to account on possible stride...

Reply:
fixed.